zilliqa/src/api/eth.rs

```diff
@@ -1039,10 +1039,114 @@ fn blob_base_fee(_params: Params, _node: &Arc<Mutex<Node>>) -> Result<()> {
+fn calculate_rewards(
+    block_gas_used: u64,
+    reward_percentiles: &[f64],
+    sorted_gas_prices: &[u128],
+) -> Vec<f64> {
+    let target_gas_values: Vec<f64> = reward_percentiles
+        .iter()
+        .map(|&percentile| percentile / 100.0 * (block_gas_used as f64))
+        .collect();
+
+    let mut rewards = vec![0.0; reward_percentiles.len()];
+    let mut sum_gas = 0.0;
+    let mut percentile_index = 0;
+
+    for gas_used in sorted_gas_prices.iter().map(|&g| g as f64) {
+        sum_gas += gas_used;
+
+        while percentile_index < reward_percentiles.len()
+            && sum_gas >= target_gas_values[percentile_index]
+        {
+            rewards[percentile_index] = gas_used;
+            percentile_index += 1;
+        }
+
+        if percentile_index >= reward_percentiles.len() {
+            break;
+        }
+    }
+
+    for i in percentile_index..reward_percentiles.len() {
+        rewards[i] = sorted_gas_prices.last().copied().unwrap_or(0) as f64;
+    }
+
+    rewards
+}
+
 /// eth_feeHistory
 /// Returns the collection of historical gas information
-fn fee_history(_params: Params, _node: &Arc<Mutex<Node>>) -> Result<()> {
-    Err(anyhow!("API method eth_feeHistory is not implemented yet"))
+fn fee_history(params: Params, node: &Arc<Mutex<Node>>) -> Result<eth::FeeHistory> {
+    let mut params = params.sequence();
+    let block_count: u64 = params.next()?;
+    let newest_block: BlockNumberOrTag = params.next()?;
+    let reward_percentiles: Option<Vec<f64>> = params.optional_next()?;
+    expect_end_of_params(&mut params, 2, 3)?;
+
+    let node = node.lock().unwrap();
+    let newest_block_number = node
+        .resolve_block_number(newest_block)?
+        .ok_or_else(|| anyhow!("block not found"))?
+        .number();
+
+    if newest_block_number < block_count {
+        return Err(anyhow!("block_count is greater than newest_block"));
+    }
+
+    let oldest_block = newest_block_number - block_count + 1;
+    let (reward, gas_used_ratio) = (oldest_block..=newest_block_number)
+        .map(|block_number| {
+            let block = node
+                .get_block(BlockNumberOrTag::Number(block_number))?
+                .ok_or_else(|| anyhow!("block not found"))?;
+
+            let reward = if let Some(reward_percentiles) = reward_percentiles.as_ref() {
+                let sorted_gas_prices = block
+                    .transactions
+                    .iter()
+                    .map(|tx_hash| {
+                        let tx = node
+                            .get_transaction_by_hash(*tx_hash)?
+                            .ok_or_else(|| anyhow!("transaction not found: {}", tx_hash))?;
+                        Ok(tx.tx.gas_price_per_evm_gas())
+                    })
+                    .collect::<Result<Vec<_>>>()?
+                    .into_iter()
+                    .sorted_unstable()
+                    .collect::<Vec<_>>();
```

Comment:
This makes two allocations. I would instead `collect` into a mutable `Vec`, then just call `Vec::sort_unstable`.